Team Profile Page

muzaffaryousaf opened the pull request:
This adds the UI for team profile page except the discussions component.

Sandbox:  http://muzaffar-team-profile.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/teams/#teams/python/python

muzaffaryousaf commented:
@explorerleslie Here is the link to sandbox --> http://muzaffar-team-profile.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/teams/#teams/python/python

muzaffaryousaf commented:
@marcotuts can you please give it a look too ?

muzaffaryousaf commented:
@symbolist  Here.

marcotuts commented:
@muzaffaryousaf I had a chance to catch up with @explorerleslie and we discussed the two options you shared. 

The UX suggestion here is actually to move toward the second option you shared in our hangout discussion, which involves an input field with selectable text on focus. Here is a screenshot of what this looks like currently as implemented on the Studio Files & Uploads page. 

Studio:
![image](https://cloud.githubusercontent.com/assets/2023680/9071766/2eb94f1e-3ac5-11e5-9ec2-8677ba2829cc.png)

I don't know if any of this Studio code can be easily reused or generalized, but it is a useful pattern to emulate for this section.

I dropped in the input field here as an example screenshot, and wanted to tag @catong here in case she has any suggestions on the copy. We used to have the words "Copy Link" as supportive text in the button link, but without this we may want to slightly adjust the header or subtitle for the usage of the input field to be clear. 
Example: 
![image](https://cloud.githubusercontent.com/assets/2023680/9072036/b925035e-3ac6-11e5-8bda-453b0ede4480.png)

catong commented:
For the help text under the URL field, I suggest:
"Send this link to friends so that they can join this team too."
I left out the word "copy" in the interests of length, but an alternative is "Copy and send this link to friends so that they can join this team too." @marcotuts @muzaffaryousaf @explorerleslie Thoughts?

explorerleslie commented:
@catong I like your suggested text.

marcotuts commented:
+1 to the copy here. thanks @catong!

muzaffaryousaf commented:
@marcotuts Thanks for your feedback. I think this makes more sense than a prompt window.

muzaffaryousaf commented:
@marcotuts @catong @explorerleslie Can you please review this PR. The sandbox is available on this url --> https://muzaffar-team-profile.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/teams

explorerleslie commented:
@muzaffaryousaf I can't seem to get the "Leave Team" link to work. I clicked it several times while logged in as audit, and it didn't do anything.

muzaffaryousaf commented:
@explorerleslie There is a separate ticket for Leave Team and Join Team. This is only read only view.

explorerleslie commented:
@muzaffaryousaf oh, right. Ok, looks good but I'll wait to give thumbs until Marco and Mark have done reviews.

muzaffaryousaf commented:
@cptvitamin Can you please give it a look ? This adds only the right panel for teams detail. Thanks.

cptvitamin commented on lms/djangoapps/teams/static/teams/templates/team-member.underscore:
```diff
@@ -0,0 +1,8 @@
+<span class="team-member">
+    <a class="member-profile" href="<%= memberProfileUrl %>">
+        <p><%- gettext(username) %></p>
+        <img class="image-url" src="<%= imageUrl %>" alt="<%- gettext(username) %>" />
```
The alt text here is redundant since the exact same text precedes it.  I would make the alt text here, the translatable string "profile page" so the screen reader reads  "@muzaffaryousaf profile page".  Currently, it would read "muzaffaryousaf muzaffaryousaf"

muzaffaryousaf replied:
Make sense.

symbolist replied:
`profile-image` will be a better class name. And again `username` should not be in `gettext`.

cptvitamin commented on lms/djangoapps/teams/static/teams/templates/team-member.underscore:
```diff
@@ -0,0 +1,8 @@
+        <p><%- gettext(username) %></p>
+        <img class="image-url" src="<%= imageUrl %>" alt="<%- gettext(username) %>" />
+    </a>
+    <span class="sr"><%- gettext("Click to visit profile for this user") %></span>
```
This screen reader only text is not necessary.  It's not programmatically linked to the control, so it won't be read along with it.  To effectively pull this off, you would need an ID on the span and then refer to the ID using aria-describedby on the profile page link. I think its too much effort for what its worth.

muzaffaryousaf replied:
Removing it.

cptvitamin commented on lms/djangoapps/teams/static/teams/templates/team-profile.underscore:
```diff
@@ -4,4 +4,77 @@
+        <h4><%- gettext("Team Details") %></h4>
+        <% if (isMember) { %>
+            <div class="team-member">
+                <span><%- gettext("You are a member of this team.") %></span>
```
should this be a ```<p>``` instead of ```<span>``` ?

cptvitamin commented on lms/djangoapps/teams/static/teams/templates/team-profile.underscore:
```diff
@@ -4,4 +4,77 @@
+
+        <% if (isMember) { %>
+            <div class="leave-team">
+                <a href="#"><%- gettext("Leave Team") %></a>
```
Can we make this a ```<button>``` ?

muzaffaryousaf replied:
We need link here not a button.

cptvitamin commented on lms/djangoapps/teams/static/teams/templates/team-profile.underscore:
```diff
@@ -4,4 +4,77 @@
+        <% if (isMember) { %>
+            <div class="leave-team">
+                <a href="#"><%- gettext("Leave Team") %></a>
+                <span class="sr"><%- gettext("Click this to leave the team") %></span>
```
This screen reader text is unnecessary as its clear what control does.

cptvitamin commented on lms/djangoapps/teams/static/teams/templates/team-profile.underscore:
```diff
@@ -4,4 +4,77 @@
+                <h4 class="invites-header"><%- gettext("Invite Others") %></h4>
+                <% if (hasCapacity) { %>
+                    <span class="sr"><%- gettext("Copy this url to invite others") %></span>
+                    <input type="text" class="invite-link-input" value="<%= inviteLink %>" readonly>
```
I like the idea of this screen reader only text, but something tells me its unnecessary.  It's very close to the same message that is displayed visually.  The same text could be used.  What we need to do to actually make this effective is to programmatically associate it with the input.    Put an ID attribute on the "invite-text" below and then add aria-describedby="[IDref]" on the ```<input>``` that way when a screen reader user moves focus to the control, they will get the associated help text.  otherwise, they will never hear it.

muzaffaryousaf replied:
Gotcha thanks.

symbolist replied:
No need for `-input` in the class name.

marcotuts commented on lms/static/sass/views/_teams.scss:
```diff
@@ -346,6 +346,106 @@
 
+  .team-profile {
+    .discussion-module {
+      width: 60%;
```
it would be great to actually echo the same primary / secondary styling and even naming convention that we use further up for the page. (lined 50 or so of this file) 

1.) This would mean we could wrap .discussion-module in something like .page-content-main {} and then .team-profile would be wrapped in .page-content-secondary {}. 

2.) In addition it would be great to reuse the flex-grid rules used above instead of directly defining width: 60% / 30%. 

___


Here is a copy / paste of the styling for the header area to convey how we might echo the use of flex-grid as well as the naming convention. 

.page-header.has-secondary {
 
       .page-header-main {
         display: inline-block;
         width: flex-grid(8,12);
       }
 
       .page-header-secondary {
         display: inline-block;
         width: flex-grid(4,12);
         @include text-align(right);
       }
   }

marcotuts commented on lms/static/sass/views/_teams.scss:
```diff
@@ -346,6 +346,106 @@
+      @include float(left);
+      @include margin($baseline, 0, $baseline, $baseline + $baseline/2);
+
+      div, h4 {
```
It would be great to avoid element styling here by using something like .team-detail-header perhaps?

marcotuts commented on lms/static/sass/views/_teams.scss:
```diff
@@ -346,6 +346,106 @@
+      div, h4 {
+        margin-bottom: ($baseline/4);
+      }
+      @extend %t-copy-sub1;
```
If this belongs to team-detail it would be good to move up to the earlier list of properties. currently it is split by the div / h4 styling.

marcotuts commented on lms/static/sass/views/_teams.scss:
```diff
@@ -346,6 +346,106 @@
+        position: relative;
+        display: inline-block;
+
+        p {
```
It would also be good to avoid the element styling here. Is this <p> tag a single member block / element?

marcotuts replied:
Ah no, it looks like this is handling the tooltip styling is that correct? I think we already define these styles elsewhere, but I could be wrong. I didn't see a data-tooltip attribute in the html so I thought I'd check.

muzaffaryousaf replied:
@marcotuts We are using the same tooltip in sequence nav bar in courseware [here](https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/css/sequence/display.scss#L189) 
Maybe we can put this sass to some common place to re use but i have no idea what will be the impact as this is in xmodule. Any suggestions ?

muzaffaryousaf replied:
@marcotuts FYI, We have another branch for Joining Team and we reused the same tooltip sass for join button by making it mixin but we put that in teams.sass.

marcotuts replied:
Ah ok thanks for the note here. thanks for clarifying!

catong commented on lms/djangoapps/teams/static/teams/templates/team-profile.underscore:
```diff
@@ -4,4 +4,77 @@
+            </div>
+        <% } %>
+        <div class="team-members">
+            <span class="sr"><%- gettext("Team members profile") %></span>
```
Should this be "Team member profiles" instead? (i.e. "profiles" is the plural word)

muzaffaryousaf replied:
Copy that.

catong commented on lms/djangoapps/teams/static/teams/templates/team-profile.underscore:
```diff
@@ -4,4 +4,77 @@
+                    </span>
+                <% } else {  %>
+                    <span class="invite-text">
+                        <%- gettext("No invites available - this team is full") %>
```
"No invitations are available. This team is full."

catong commented on lms/djangoapps/teams/static/teams/templates/team-profile.underscore:
```diff
@@ -4,4 +4,77 @@
+        </div>
+
+        <div class="team-country">
+            <span class="sr"><%- gettext("Team location") %></span>
```
I'm not sure it will always be the actual location of the team. 
"The country that team members primarily identify with." might be more accurate, as this is how we describe it in the Create Team page.  @explorerleslie Thoughts?

explorerleslie replied:
agreed @catong

catong commented on lms/djangoapps/teams/static/teams/templates/team-profile.underscore:
```diff
@@ -4,4 +4,77 @@
+            </span>
+        </div>
+        <div class="team-language">
+            <span class="sr"><%- gettext("Team language") %></span>
```
"The language that team members primarily use to communicate with each other."

catong commented:
@muzaffaryousaf Thanks, I gave some suggestions for text, including some SR only text.

muzaffaryousaf commented:
@symbolist Can you please give it a rough look while i write test cases. Thanks.

symbolist commented on lms/djangoapps/teams/static/teams/js/views/team_profile.js:
```diff
@@ -18,13 +27,67 @@
+                renderTeamMembers: function() {
+                    var view = this;
+                    _.each(this.memberships, function(data) {
+                        $.ajax({
```
Making separate requests just to get the profile data of members is clearly not the right way to go about it. I looked at the `TeamsDetailView` and it seems to be inlining the accounts data of the members. I have a feeling this should be the profile data of members. Can you ask the Cambridge team about this?

symbolist commented on lms/static/sass/views/_teams.scss:
```diff
@@ -346,6 +346,106 @@
+        }
+      }
+
+    //TODO remove these divider once the Create team merge in master
```
Can we remove these now?

symbolist commented on lms/djangoapps/teams/views.py:
```diff
@@ -100,6 +100,7 @@ def get(self, request, course_id):
             "teams_url": reverse('teams_list', request=request),
             "languages": settings.ALL_LANGUAGES,
             "countries": list(countries),
+            "current_username": request.user.username
```
I think `request_username` will convey better what this variable is.

symbolist commented on lms/djangoapps/teams/static/teams/templates/team-profile.underscore:
```diff
@@ -4,4 +4,77 @@
+            <i class="u-field-icon icon fa fa-globe fa-fw" aria-hidden="true"></i>
+            <span>
+                <%=
+                    interpolate_text(
```
Why the need to use `interpolate`? Will `<%- gettext(country) %>` not work here?

symbolist commented on lms/djangoapps/teams/static/teams/templates/team-profile.underscore:
```diff
@@ -4,4 +4,77 @@
+
+            <div class="divider-lv1"></div>
+
+            <div class="invite-team">
```
How about `team-invites` to be consistent with other class names?

symbolist replied:
Just to make the pluralization consistent lets go with `team-invite` and `invite-header`.

symbolist commented on lms/djangoapps/teams/static/teams/templates/team-member.underscore:
```diff
@@ -0,0 +1,8 @@
+
+<span class="team-member">
+    <a class="member-profile" href="<%= memberProfileUrl %>">
+        <p><%- gettext(username) %></p>
```
No need for `gettext` here.

symbolist commented on lms/djangoapps/teams/static/teams/js/views/team_profile.js:
```diff
@@ -18,13 +27,67 @@
+                },
+                isUserMemberOfTeam: function() {
+                    var self = this;
+                    var member = _.find(this.memberships, function (member) {
```
The parameter of the function should be `membership` and not `member`.

symbolist commented on lms/djangoapps/teams/static/teams/templates/team-profile.underscore:
```diff
@@ -4,4 +4,77 @@
+    <div class="team-detail">
+        <h4><%- gettext("Team Details") %></h4>
+        <% if (isMember) { %>
+            <div class="team-member">
```
How about `team-membership-status`?

symbolist commented on lms/djangoapps/teams/static/teams/js/views/team_profile.js:
```diff
@@ -18,13 +27,67 @@
+                                '%(member_count)s / %(max_member_count)s Members',
+                                this.maxTeamSize
+                            ),
+                            {member_count: this.memberships.length, max_member_count: this.maxTeamSize}, true
```
Can we make this a function and move it to a helper file? This code is in `TeamMembershipView` as well.

symbolist commented on lms/djangoapps/teams/static/teams/js/views/team_profile.js:
```diff
@@ -18,13 +27,67 @@
+                        ),
+                        isMember: this.isUserMemberOfTeam(),
+                        hasCapacity: this.memberships.length < this.maxTeamSize,
+                        inviteLink: window.location +'?invite=true'
```
This may not be deterministic. Is there a way to construct this url instead?

symbolist commented:
@muzaffaryousaf I am done with my first pass. We need to figure out the profile data issue as soon as possible. Otherwise the code looks good.

muzaffaryousaf commented:
@cptvitamin @catong I've addressed your comments, please give it a final look. Thanks.

catong commented:
@muzaffaryousaf Thank you! :+1:  for doc strings

explorerleslie commented:
A few minor comments: 

@muzaffaryousaf country names and languages should both be the full versions of the names. Right now country looks like it's using abbreviations (UK) and it should use full name. 

@marcotuts do we need that bottom separator line after the invite link? Looks a little weird to my eye because I'm expecting something else to appear below it, but I'll leave you to make the final call. 

@marcotuts @muzaffaryousaf @andy-armstrong I feel like the widths of the two columns needs adjustments. The right sidebar is too wide. And the two columns don't fill the entire center section. Where should we adjust this? I know this is probably off because we've been working on this in two PRs.

muzaffaryousaf commented:
@explorerleslie I've fixed the countries and languages.
Also, trying to fix the two columns. I will update the sandbox once i'm done with the changes.

explorerleslie commented:
@muzaffaryousaf one more thing: when language and country aren't specified, we shouldn't show those indicators at all on the right hand side of the teams page. Right now just hte icons hang out there. 

![image](https://cloud.githubusercontent.com/assets/5643426/9113508/072900fc-3c23-11e5-9374-37c793ba1bde.png)

muzaffaryousaf commented:
@explorerleslie all set for country and language. Thanks.

marcotuts commented on lms/static/sass/views/_teams.scss:
```diff
@@ -346,6 +346,97 @@
+      }
+
+      .leave-team, .invites-header, .invite-text {
+        margin-bottom: ($baseline/2);
```
NIT: I would switch to $baseline for a bit more spacing here.

marcotuts commented on lms/static/sass/views/_teams.scss:
```diff
@@ -346,6 +346,97 @@
+      }
+
+      .invite-team {
+        margin-top: ($baseline/2);
```
NIT: I would switch this to $baseline for a bit more spacing here.

marcotuts commented on lms/djangoapps/teams/static/teams/templates/team-profile.underscore:
```diff
@@ -0,0 +1,68 @@
+                    <% } %>
+                </div>
+
+                <div class="divider-lv1"></div>
```
NIT: +1 to @explorerleslie 's suggestion to drop this divider as it is the last section on the right sidebar.

marcotuts commented:
One last note / follow-up based on @explorerleslie's note about the width of the discussions area. 

We've set the main content area to 60% and the sidebar to 30% here (with 10% off to the right of the sidebar unused currently). Whether it is in this PR or a future one, using flex-grid(8 of 12) for the left and flex-grid(4 of 12) for the right side shifts the ratio a bit closer to what I think we're expecting.

Reference image: 
![image](https://cloud.githubusercontent.com/assets/2023680/9114924/6299a098-3c2a-11e5-8725-968a6f15b838.png)

muzaffaryousaf commented:
@marcotuts as per your note about the flex-grid i've set `flex-grid(7,12)` for main content and `flex-grid(4,12)` for secondary page content which you can see on the sandbox. I've tried to put `flex-grid(8,12)` for main content but in that case the right panel comes down because of less space. The only way to set them side by side was removing the margin between these two containers but that will not leave any space between them which looks bad. Can you suggest something to adjust these containers. ? Thanks...

marcotuts commented:
@muzaffaryousaf I can try this out locally if that helps, but my quick guess after looking at the styling is that the margin: $baseline rule on the right sidebar was actually adding 40px of width to the page on top of the flex-grid defined width. If that is changed to padding it should work ok. 

In a related note, it might also help to remove .team-profile as a layer in the HTML. With the new page-content-main and page-content-secondary, this div doesn't seem critical anymore.

muzaffaryousaf commented:
@marcotuts I've addressed your comments, can you please give it a final look ? Thanks.
For now the profile image urls will be same for all users because of a bug in API, Cambridge is working on that, you can ignore them. Thanks.

muzaffaryousaf commented:
@marcotuts @cptvitamin  I've updated the sandbox with latest changes. Please have look.
Thanks.

marcotuts commented:
@muzaffaryousaf I think what you have currently is an improvement ratio-wise to what we had before, so I'm ok with this as is. We can revisit the styling / layout in a follow-on story, especially since we hope at some point to replace flex-grid with Suzy anyways. (cc'ed @explorerleslie  on this layout update since you originally noted it.)

:+1: from me. 

(One note - possibly a separate PR but I still see an error when I try to expand the discussion. "This thread has been deleted," aka our generic error message. I don't think that's necessarily a part of this PR, but I thought I'd note it in passing.)

muzaffaryousaf commented:
@explorerleslie Please ignore language & country names for now. I'm waiting on this [PR](https://github.com/edx/edx-platform/pull/9206) to get merge. Also, ignore the profile images, they are same. Cambridge is working on the profile images url bug, i'm blocked on them for profile image of each user.

muzaffaryousaf commented:
Closing this in favour of this https://github.com/edx/edx-platform/pull/9228
